File: apps/web/src/offline/sync-coordinator.test.ts
import assert from "node:assert/strict";
import test from "node:test";
import { MemorySyncRepository } from "./memory-sync-repository.ts";
import { SyncCoordinator, type SyncTransport } from "./sync-coordinator.ts";
import { SYNC_PROTOCOL_VERSION, type SyncCommandEnvelope } from "./sync-contract.ts";

const partition = { tenantId: "tenant-a", deviceId: "device-a" } as const;
const command: SyncCommandEnvelope = {
  version: SYNC_PROTOCOL_VERSION,
  commandId: "11111111-1111-4111-8111-111111111111",
  type: "sales.finalize.v1",
  occurredAt: "2026-08-08T10:00:00Z",
  payload: { saleId: "sale-1" },
};

test("reconnect pulls gapped pages, flushes the outbox, and pulls command changes", async () => {
  const repository = new MemorySyncRepository();
  await repository.enqueue(partition, command);
  const requests: string[] = [];
  let pullCount = 0;
  const transport: SyncTransport = {
    pull: async (_partition, cursor) => {
      requests.push(`pull:${cursor}`);
      pullCount += 1;
      if (pullCount === 1) {
        return {
          version: "1", cursor: 4, hasMore: true,
          changes: [{ cursor: 4, entityType: "products", entityId: "p1", operation: "upsert", payload: { name: "Milk" }, occurredAt: "2026-08-08T10:00:00Z" }],
        };
      }
      if (pullCount === 2) {
        return {
          version: "1", cursor: 5, hasMore: false,
          changes: [{ cursor: 5, entityType: "products", entityId: "p2", operation: "upsert", payload: { name: "Bread" }, occurredAt: "2026-08-08T10:00:01Z" }],
        };
      }
      return {
        version: "1", cursor: 6, hasMore: false,
        changes: [{ cursor: 6, entityType: "sales", entityId: "sale-1", operation: "upsert", payload: { status: "finalized" }, occurredAt: "2026-08-08T10:00:02Z" }],
      };
    },
    submit: async (_partition, submitted) => {
      requests.push(`submit:${submitted.commandId}`);
      return { commandId: submitted.commandId, status: "applied", attempts: 1 };
    },
  };

  const result = await new SyncCoordinator(repository, transport).reconnect(partition);
  assert.deepEqual(requests, [
    "pull:0", "pull:4", "submit:11111111-1111-4111-8111-111111111111", "pull:5",
  ]);
  assert.deepEqual(result, {
    pagesPulled: 3, changesApplied: 3, commandsSubmitted: 1,
    commandsApplied: 1, commandsRejected: 0, commandsConflicted: 0, commandsRetryPending: 0,
  });
  assert.equal((await repository.cursor(partition)), 6);
  assert.deepEqual((await repository.projection(partition, "sales"))[0]?.value, { status: "finalized" });
  assert.equal((await repository.outbox(partition))[0]?.state, "applied");
});

test("retry_pending returns the command to pending without losing its id", async () => {
  const repository = new MemorySyncRepository();
  await repository.enqueue(partition, command);
  const transport: SyncTransport = {
    pull: async () => ({ version: "1", cursor: 0, hasMore: false, changes: [] }),
    submit: async (_partition, submitted) => ({ commandId: submitted.commandId, status: "retry_pending", attempts: 1, resultCode: "worker_busy" }),
  };
  const result = await new SyncCoordinator(repository, transport).reconnect(partition);
  assert.equal(result.commandsRetryPending, 1);
  assert.equal((await repository.outbox(partition))[0]?.state, "pending");
  assert.equal((await repository.outbox(partition))[0]?.envelope.commandId, command.commandId);
});

test("a transport interruption recovers claimed commands for a later reconnect", async () => {
  const repository = new MemorySyncRepository();
  await repository.enqueue(partition, command);
  const transport: SyncTransport = {
    pull: async () => ({ version: "1", cursor: 0, hasMore: false, changes: [] }),
    submit: async () => { throw new Error("offline"); },
  };
  await assert.rejects(new SyncCoordinator(repository, transport).reconnect(partition), /offline/);
  assert.equal((await repository.outbox(partition))[0]?.state, "pending");
  assert.equal((await repository.outbox(partition))[0]?.attempts, 1);
});

File: apps/mobile/test/sync_coordinator_test.dart
import 'package:nova_mobile/nova_mobile.dart';
import 'package:test/test.dart';

final class FakeTransport implements SyncTransport {
  FakeTransport(this.pages, this.receipts);

  final List<SyncChangePage> pages;
  final List<SyncCommandReceipt> receipts;
  final submitted = <String>[];
  int pageIndex = 0;
  int receiptIndex = 0;
  Object? submitError;

  @override
  Future<SyncChangePage> pull(SyncPartition partition, int cursor, {int limit = 500}) async => pages[pageIndex++];

  @override
  Future<SyncCommandReceipt> submit(SyncPartition partition, OutboxCommand command) async {
    submitted.add(command.commandId);
    final error = submitError;
    if (error != null) throw error!;
    return receipts[receiptIndex++];
  }
}

void main() {
  const partition = SyncPartition(tenantId: 'tenant-a', deviceId: 'device-a');

  ProjectionChange change(int cursor, String id) => ProjectionChange(
        cursor: cursor,
        entityType: 'product',
        entityId: id,
        operation: 'upsert',
        payload: {'name': id},
        occurredAt: DateTime.utc(2026, 8, 18),
      );

  OutboxCommand command() => OutboxCommand(
        commandId: 'command-1',
        type: 'sales.finalize.v1',
        payload: {'saleId': 'sale-1'},
        occurredAt: DateTime.utc(2026, 8, 18),
      );

  SyncCommandReceipt receipt(SyncReceiptStatus status) => SyncCommandReceipt(
        commandId: 'command-1',
        status: status,
        attempts: 1,
      );

  SyncChangePage page(List<ProjectionChange> changes, bool hasMore) => SyncChangePage(
        cursor: changes.isEmpty ? 0 : changes.last.cursor,
        changes: changes,
        hasMore: hasMore,
      );

  test('reconnect pulls, submits, and performs a final change pull', () async {
    final repository = MemorySyncRepository();
    await repository.enqueue(partition, command());
    final transport = FakeTransport(
      [page([change(2, 'p1')], false), page([change(4, 'sale-1')], false)],
      [receipt(SyncReceiptStatus.applied)],
    );

    final result = await SyncCoordinator(repository, transport).reconnect(partition);

    expect(result.pagesPulled, 2);
    expect(result.changesApplied, 2);
    expect(result.commandsSubmitted, 1);
    expect(result.commandsApplied, 1);
    expect(await repository.cursor(partition), 4);
    expect(await repository.projection(partition, 'product', 'p1'), {'name': 'p1'});
    expect(await repository.pending(partition), isEmpty);
  });

  test('retry-pending returns the command to pending without spinning', () async {
    final repository = MemorySyncRepository();
    await repository.enqueue(partition, command());
    final transport = FakeTransport(
      [page([], false)],
      [receipt(SyncReceiptStatus.retryPending)],
    );

    final result = await SyncCoordinator(repository, transport).reconnect(partition);

    expect(result.commandsRetryPending, 1);
    expect(transport.submitted, ['command-1']);
    expect((await repository.pending(partition)).single.status, OutboxStatus.pending);
  });

  test('transport failure recovers a claimed command for a later pass', () async {
    final repository = MemorySyncRepository();
    await repository.enqueue(partition, command());
    final transport = FakeTransport([page([], false)], const []);
    transport.submitError = StateError('offline');

    await expectLater(
      SyncCoordinator(repository, transport).reconnect(partition),
      throwsA(isA<StateError>()),
    );
    expect((await repository.pending(partition)).single.status, OutboxStatus.pending);
    expect((await repository.pending(partition)).single.attempts, 1);
  });

  test('rejects a page that claims more work without cursor progress', () async {
    final repository = MemorySyncRepository();
    final transport = FakeTransport([page([], true)], const []);

    await expectLater(
      SyncCoordinator(repository, transport).reconnect(partition),
      throwsA(isA<SyncRepositoryException>()),
    );
    expect((await repository.cursor(partition)), 0);
  });
}
